examples/flax/language-modeling/run_clm_flax.py

```diff
@@ -526,15 +537,28 @@ def compute_loss(params):
             return loss
 
         grad_fn = jax.value_and_grad(compute_loss)
-        loss, grad = grad_fn(state.params)
-        grad = jax.lax.pmean(grad, "batch")
+        loss, grads = grad_fn(state.params)
+        grad_accum = jax.tree_multimap(lambda x, y: x + y, grads, state.grad_accum)
+
+        def update_fn():
+            grads = jax.tree_map(lambda x: x / training_args.gradient_accumulation_steps, grad_accum)
+            grads = jax.lax.pmean(grads, "batch")
+            new_state = state.apply_gradients(
+                grads=grads, grad_accum=jax.tree_map(jnp.zeros_like, grads), optimizer_step=state.optimizer_step + 1
+            )
+            return new_state
 
-        new_state = state.apply_gradients(grads=grad, dropout_rng=new_dropout_rng)
+        new_state = jax.lax.cond(
+            state.step % training_args.gradient_accumulation_steps == 0,
```

Comment:
Does this make sure gradients are also accumulated for the first step? 
Is state.step = 1 the very first time?

Reply:
I think this implies that the very first time this function is called that `state.step` is actually 0 which would mean that instead of accumulating the gradient an optimization step is done no?

Reply:
Yes it should probably be testing `state.step+1` against the `training_args.gradient_accumulation_steps`.